cpp/src/arrow/compute/exec/data_holder_node.cc

```diff
@@ -0,0 +1,222 @@
+  }
+  AsyncGenerator<std::shared_ptr<DataHolder>> generator() { return gen_; }
+
+ public:
+  ExecContext* context_;
+  PushGenerator<std::shared_ptr<DataHolder>> gen_;
+  PushGenerator<std::shared_ptr<DataHolder>>::Producer producer_;
+};
+
+class DataHolderNode : public ExecNode {
+ public:
+  DataHolderNode(ExecPlan* plan, NodeVector inputs, std::vector<std::string> input_labels,
+                 std::shared_ptr<Schema> output_schema, int num_outputs)
+      : ExecNode(plan, std::move(inputs), input_labels, std::move(output_schema),
+                 /*num_outputs=*/num_outputs) {
+    executor_ = plan->exec_context()->executor();
+
+    data_holder_manager_ =
+        ::arrow::internal::make_unique<DataHolderManager>(plan->exec_context());
+
+    auto status = task_group_.AddTask([this]() -> Result<Future<>> {
+      ARROW_DCHECK(executor_ != nullptr);
+      return executor_->Submit(this->stop_source_.token(), [this] {
+        auto generator = this->data_holder_manager_->generator();
+        auto iterator = MakeGeneratorIterator(std::move(generator));
+        while (true) {
+          ARROW_ASSIGN_OR_RAISE(auto result, iterator.Next());
+          if (IsIterationEnd(result)) {
+            break;
+          }
+          ARROW_ASSIGN_OR_RAISE(ExecBatch batch, result->Get());
+          this->outputs_[0]->InputReceived(this, batch);
+        }
+        return Status::OK();
+      });
+    });
+    if (!status.ok()) {
+      if (input_counter_.Cancel()) {
+        this->Finish(status);
+      }
+      inputs_[0]->StopProducing(this);
+    }
+  }
+
+  void ErrorReceived(ExecNode* input, Status error) override {
+    DCHECK_EQ(input, inputs_[0]);
+    outputs_[0]->ErrorReceived(this, std::move(error));
+  }
+
+  void InputFinished(ExecNode* input, int total_batches) override {
+    DCHECK_EQ(input, inputs_[0]);
+    outputs_[0]->InputFinished(this, total_batches);
+    if (input_counter_.SetTotal(total_batches)) {
+      this->Finish();
+    }
+  }
+
+  static Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                                const ExecNodeOptions& options) {
+    auto schema = inputs[0]->output_schema();
+    return plan->EmplaceNode<DataHolderNode>(plan, std::move(inputs),
+                                             std::vector<std::string>{"target"},
+                                             std::move(schema), /*num_outputs=*/1);
+  }
+
+  const char* kind_name() const override { return "DataHolderNode"; }
+
+  void InputReceived(ExecNode* input, ExecBatch batch) override {
+    if (finished_.is_finished()) {
+      return;
+    }
+    auto status = task_group_.AddTask([this, batch]() -> Result<Future<>> {
+      return this->executor_->Submit(this->stop_source_.token(), [this, batch]() {
+        auto pool = this->plan()->exec_context()->memory_pool();
+        ARROW_ASSIGN_OR_RAISE(auto record_batch,
+                              batch.ToRecordBatch(this->output_schema(), pool));
```

Comment:
Converting every `ExecBatch` to `RecordBatch` sort of defeats the purpose of having `ExecBatch` in the first place.  Ideally this would only have to happen if we needed to store to disk.